// ant-design-vue/lib/date-picker/interface.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WeekPickerProps = exports.RangePickerProps = exports.MonthPickerProps = exports.DatePickerProps = exports.SinglePickerProps = exports.PickerProps = void 0;

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

var _momentUtil = require("../_util/moment-util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var PickerProps = function PickerProps() {
  return {
    name: _vueTypes.default.string,
    transitionName: _vueTypes.default.string,
    prefixCls: _vueTypes.default.string,
    inputPrefixCls: _vueTypes.default.string,
    format: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.array, _vueTypes.default.func]),
    disabled: _vueTypes.default.bool,
    allowClear: _vueTypes.default.bool,
    suffixIcon: _vueTypes.default.any,
    popupStyle: _vueTypes.default.object,
    dropdownClassName: _vueTypes.default.string,
    locale: _vueTypes.default.any,
    localeCode: _vueTypes.default.string,
    size: _vueTypes.default.oneOf(['large', 'small', 'default']),
    getCalendarContainer: _vueTypes.default.func,
    open: _vueTypes.default.bool,
    disabledDate: _vueTypes.default.func,
    showToday: _vueTypes.default.bool,
    dateRender: _vueTypes.default.any,
    // ({current: moment.Moment, today: moment.Moment}) => vNode,
    pickerClass: _vueTypes.default.string,
    pickerInputClass: _vueTypes.default.string,
    timePicker: _vueTypes.default.any,
    autofocus: _vueTypes.default.bool,
    tagPrefixCls: _vueTypes.default.string,
    tabindex: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]),
    align: _vueTypes.default.object.def(function () {
      return {};
    }),
    inputReadOnly: _vueTypes.default.bool,
    valueFormat: _vueTypes.default.string,
    onOpenChange: _vueTypes.default.func,
    onFocus: _vueTypes.default.func,
    onBlur: _vueTypes.default.func,
    'onUpdate:value': _vueTypes.default.func,
    onMouseenter: _vueTypes.default.func,
    onMouseleave: _vueTypes.default.func
  };
};

exports.PickerProps = PickerProps;

var SinglePickerProps = function SinglePickerProps() {
  return {
    value: _momentUtil.TimeType,
    defaultValue: _momentUtil.TimeType,
    defaultPickerValue: _momentUtil.TimeType,
    renderExtraFooter: _vueTypes.default.any,
    placeholder: _vueTypes.default.string,
    onChange: _vueTypes.default.func
  };
};

exports.SinglePickerProps = SinglePickerProps;

var DatePickerProps = function DatePickerProps() {
  return _objectSpread(_objectSpread(_objectSpread({}, PickerProps()), SinglePickerProps()), {}, {
    showTime: _vueTypes.default.oneOfType([_vueTypes.default.object, _vueTypes.default.bool]),
    open: _vueTypes.default.bool,
    disabledTime: _vueTypes.default.func,
    mode: _vueTypes.default.oneOf(['time', 'date', 'month', 'year']),
    onOpenChange: _vueTypes.default.func,
    onPanelChange: _vueTypes.default.func,
    onOk: _vueTypes.default.func
  });
};

exports.DatePickerProps = DatePickerProps;

var MonthPickerProps = function MonthPickerProps() {
  return _objectSpread(_objectSpread(_objectSpread({}, PickerProps()), SinglePickerProps()), {}, {
    placeholder: _vueTypes.default.string,
    monthCellContentRender: _vueTypes.default.func
  });
}; // export const RangePickerPresetRange = PropTypes.oneOfType([TimesType, PropTypes.func])


exports.MonthPickerProps = MonthPickerProps;

var RangePickerProps = function RangePickerProps() {
  return _objectSpread(_objectSpread({}, PickerProps()), {}, {
    tagPrefixCls: _vueTypes.default.string,
    value: _momentUtil.TimesType,
    defaultValue: _momentUtil.TimesType,
    defaultPickerValue: _momentUtil.TimesType,
    timePicker: _vueTypes.default.any,
    showTime: _vueTypes.default.oneOfType([_vueTypes.default.object, _vueTypes.default.bool]),
    ranges: _vueTypes.default.object,
    placeholder: _vueTypes.default.arrayOf(String),
    mode: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.arrayOf(String)]),
    separator: _vueTypes.default.any,
    disabledTime: _vueTypes.default.func,
    showToday: _vueTypes.default.bool,
    renderExtraFooter: _vueTypes.default.any,
    onChange: _vueTypes.default.func,
    onCalendarChange: _vueTypes.default.func,
    onOk: _vueTypes.default.func,
    onPanelChange: _vueTypes.default.func,
    onMouseenter: _vueTypes.default.func,
    onMouseleave: _vueTypes.default.func
  });
};

exports.RangePickerProps = RangePickerProps;

var WeekPickerProps = function WeekPickerProps() {
  return _objectSpread(_objectSpread(_objectSpread({}, PickerProps()), SinglePickerProps()), {}, {
    placeholder: _vueTypes.default.string
  });
}; // export interface DatePickerDecorator extends React.ClassicComponentClass<DatePickerProps> {
//   RangePicker: React.ClassicComponentClass<RangePickerProps>;
//   MonthPicker: React.ClassicComponentClass<MonthPickerProps>;
//   WeekPicker: React.ClassicComponentClass<WeexPickerProps>;
// }


exports.WeekPickerProps = WeekPickerProps;
