// ant-design-vue/lib/vc-select/PropTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectPropTypes = void 0;

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SelectPropTypes = {
  defaultActiveFirstOption: _vueTypes.default.bool,
  multiple: _vueTypes.default.bool,
  filterOption: _vueTypes.default.any,
  // children: PropTypes.any,
  showSearch: _vueTypes.default.bool,
  disabled: _vueTypes.default.bool,
  allowClear: _vueTypes.default.bool,
  showArrow: _vueTypes.default.bool,
  tags: _vueTypes.default.bool,
  prefixCls: _vueTypes.default.string,
  // className: PropTypes.string,
  transitionName: _vueTypes.default.string,
  optionLabelProp: _vueTypes.default.string,
  optionFilterProp: _vueTypes.default.string,
  animation: _vueTypes.default.string,
  choiceTransitionName: _vueTypes.default.string,
  open: _vueTypes.default.bool,
  defaultOpen: _vueTypes.default.bool,
  // onChange: PropTypes.func,
  // onBlur: PropTypes.func,
  // onFocus: PropTypes.func,
  // onSelect: PropTypes.func,
  // onSearch: PropTypes.func,
  // onPopupScroll: PropTypes.func,
  // onMouseEnter: PropTypes.func,
  // onMouseLeave: PropTypes.func,
  // onInputKeyDown: PropTypes.func,
  placeholder: _vueTypes.default.any,
  // onDeselect: PropTypes.func,
  labelInValue: _vueTypes.default.bool,
  loading: _vueTypes.default.bool,
  value: _vueTypes.default.any,
  defaultValue: _vueTypes.default.any,
  dropdownStyle: _vueTypes.default.object,
  dropdownClassName: _vueTypes.default.string,
  maxTagTextLength: _vueTypes.default.number,
  maxTagCount: _vueTypes.default.number,
  maxTagPlaceholder: _vueTypes.default.any,
  tokenSeparators: _vueTypes.default.arrayOf(_vueTypes.default.string),
  getInputElement: _vueTypes.default.func,
  showAction: _vueTypes.default.arrayOf(_vueTypes.default.string),
  autofocus: _vueTypes.default.bool,
  getPopupContainer: _vueTypes.default.func,
  clearIcon: _vueTypes.default.any,
  inputIcon: _vueTypes.default.any,
  removeIcon: _vueTypes.default.any,
  menuItemSelectedIcon: _vueTypes.default.any,
  dropdownRender: _vueTypes.default.func,
  mode: _vueTypes.default.oneOf(['multiple', 'tags']),
  backfill: _vueTypes.default.bool,
  dropdownAlign: _vueTypes.default.any,
  dropdownMatchSelectWidth: _vueTypes.default.bool,
  dropdownMenuStyle: _vueTypes.default.object,
  notFoundContent: _vueTypes.default.oneOfType([String, Number]),
  tabindex: _vueTypes.default.oneOfType([String, Number]),
  __propsSymbol__: _vueTypes.default.any,
  children: _vueTypes.default.array
};
exports.SelectPropTypes = SelectPropTypes;
