// ant-design-vue/lib/statistic/Statistic.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.StatisticProps = void 0;

var _vue = require("vue");

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

var _propsUtil = require("../_util/props-util");

var _configProvider = require("../config-provider");

var _Number = _interopRequireDefault(require("./Number"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var StatisticProps = {
  prefixCls: _vueTypes.default.string,
  decimalSeparator: _vueTypes.default.string,
  groupSeparator: _vueTypes.default.string,
  format: _vueTypes.default.string,
  value: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number, _vueTypes.default.object]),
  valueStyle: _vueTypes.default.any,
  valueRender: _vueTypes.default.any,
  formatter: _vueTypes.default.any,
  precision: _vueTypes.default.number,
  prefix: _vueTypes.default.any,
  suffix: _vueTypes.default.any,
  title: _vueTypes.default.any,
  onFinish: _vueTypes.default.func
};
exports.StatisticProps = StatisticProps;
var _default = {
  name: 'AStatistic',
  props: (0, _propsUtil.initDefaultProps)(StatisticProps, {
    decimalSeparator: '.',
    groupSeparator: ','
  }),
  setup: function setup() {
    return {
      configProvider: (0, _vue.inject)('configProvider', _configProvider.ConfigConsumerProps)
    };
  },
  render: function render() {
    var _this$$props = this.$props,
        customizePrefixCls = _this$$props.prefixCls,
        _this$$props$value = _this$$props.value,
        value = _this$$props$value === void 0 ? 0 : _this$$props$value,
        valueStyle = _this$$props.valueStyle,
        valueRender = _this$$props.valueRender;
    var getPrefixCls = this.configProvider.getPrefixCls;
    var prefixCls = getPrefixCls('statistic', customizePrefixCls);
    var title = (0, _propsUtil.getComponent)(this, 'title');
    var prefix = (0, _propsUtil.getComponent)(this, 'prefix');
    var suffix = (0, _propsUtil.getComponent)(this, 'suffix');
    var formatter = (0, _propsUtil.getComponent)(this, 'formatter', {}, false);

    var props = _objectSpread(_objectSpread({}, this.$props), {}, {
      prefixCls: prefixCls,
      value: value,
      formatter: formatter
    });

    var valueNode = (0, _vue.createVNode)(_Number.default, props, null);

    if (valueRender) {
      valueNode = valueRender(valueNode);
    }

    return (0, _vue.createVNode)("div", {
      "class": prefixCls
    }, [title && (0, _vue.createVNode)("div", {
      "class": "".concat(prefixCls, "-title")
    }, [title]), (0, _vue.createVNode)("div", {
      "style": valueStyle,
      "class": "".concat(prefixCls, "-content")
    }, [prefix && (0, _vue.createVNode)("span", {
      "class": "".concat(prefixCls, "-content-prefix")
    }, [prefix]), valueNode, suffix && (0, _vue.createVNode)("span", {
      "class": "".concat(prefixCls, "-content-suffix")
    }, [suffix])])]);
  }
};
exports.default = _default;
