// ant-design-vue/lib/vc-mentions/src/mentionsProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.defaultProps = exports.vcMentionsProps = exports.mentionsProps = void 0;

var _vueTypes = _interopRequireDefault(require("../../_util/vue-types"));

var _propsUtil = require("../../_util/props-util");

var _util = require("./util");

var _placement = require("./placement");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var mentionsProps = {
  autofocus: _vueTypes.default.bool,
  prefix: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.array]),
  prefixCls: _vueTypes.default.string,
  value: _vueTypes.default.string,
  defaultValue: _vueTypes.default.string,
  disabled: _vueTypes.default.bool,
  notFoundContent: _vueTypes.default.any,
  split: _vueTypes.default.string,
  transitionName: _vueTypes.default.string,
  placement: _vueTypes.default.oneOf(_placement.PlaceMent),
  character: _vueTypes.default.any,
  characterRender: _vueTypes.default.func,
  filterOption: _vueTypes.default.func,
  validateSearch: _vueTypes.default.func,
  getPopupContainer: _vueTypes.default.func
};
exports.mentionsProps = mentionsProps;

var vcMentionsProps = _objectSpread(_objectSpread({}, mentionsProps), {}, {
  children: _vueTypes.default.any
});

exports.vcMentionsProps = vcMentionsProps;
var defaultProps = {
  prefix: '@',
  split: ' ',
  validateSearch: _util.validateSearch,
  filterOption: _util.filterOption
};
exports.defaultProps = defaultProps;

var _default = (0, _propsUtil.initDefaultProps)(vcMentionsProps, defaultProps);

exports.default = _default;
