// ant-design-vue/lib/dropdown/getDropdownProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = function _default() {
  return {
    trigger: _vueTypes.default.array.def(['hover']),
    overlay: _vueTypes.default.any,
    visible: _vueTypes.default.bool,
    disabled: _vueTypes.default.bool,
    align: _vueTypes.default.object,
    getPopupContainer: _vueTypes.default.func,
    prefixCls: _vueTypes.default.string,
    transitionName: _vueTypes.default.string,
    placement: _vueTypes.default.oneOf(['topLeft', 'topCenter', 'topRight', 'bottomLeft', 'bottomCenter', 'bottomRight']),
    overlayClassName: _vueTypes.default.string,
    overlayStyle: _vueTypes.default.object,
    forceRender: _vueTypes.default.bool,
    mouseEnterDelay: _vueTypes.default.number,
    mouseLeaveDelay: _vueTypes.default.number,
    openClassName: _vueTypes.default.string,
    minOverlayWidthMatchTrigger: _vueTypes.default.bool
  };
};

exports.default = _default;
