// ant-design-vue/lib/timeline/TimelineItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TimeLineItemProps = void 0;

var _vue = require("vue");

var _classNames3 = _interopRequireDefault(require("../_util/classNames"));

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

var _propsUtil = require("../_util/props-util");

var _configProvider = require("../config-provider");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var TimeLineItemProps = {
  prefixCls: _vueTypes.default.string,
  color: _vueTypes.default.string,
  dot: _vueTypes.default.any,
  pending: _vueTypes.default.bool,
  position: _vueTypes.default.oneOf(['left', 'right', '']).def('')
};
exports.TimeLineItemProps = TimeLineItemProps;
var _default = {
  name: 'ATimelineItem',
  props: (0, _propsUtil.initDefaultProps)(TimeLineItemProps, {
    color: 'blue',
    pending: false
  }),
  setup: function setup() {
    var configProvider = (0, _vue.inject)('configProvider', _configProvider.ConfigConsumerProps);
    return {
      configProvider: configProvider
    };
  },
  render: function render() {
    var _classNames, _classNames2;

    var _getOptionProps = (0, _propsUtil.getOptionProps)(this),
        customizePrefixCls = _getOptionProps.prefixCls,
        _getOptionProps$color = _getOptionProps.color,
        color = _getOptionProps$color === void 0 ? '' : _getOptionProps$color,
        pending = _getOptionProps.pending;

    var getPrefixCls = this.configProvider.getPrefixCls;
    var prefixCls = getPrefixCls('timeline', customizePrefixCls);
    var dot = (0, _propsUtil.getComponent)(this, 'dot');
    var itemClassName = (0, _classNames3.default)((_classNames = {}, _defineProperty(_classNames, "".concat(prefixCls, "-item"), true), _defineProperty(_classNames, "".concat(prefixCls, "-item-pending"), pending), _classNames));
    var dotClassName = (0, _classNames3.default)((_classNames2 = {}, _defineProperty(_classNames2, "".concat(prefixCls, "-item-head"), true), _defineProperty(_classNames2, "".concat(prefixCls, "-item-head-custom"), dot), _defineProperty(_classNames2, "".concat(prefixCls, "-item-head-").concat(color), true), _classNames2));
    return (0, _vue.createVNode)("li", {
      "class": itemClassName
    }, [(0, _vue.createVNode)("div", {
      "class": "".concat(prefixCls, "-item-tail")
    }, null), (0, _vue.createVNode)("div", {
      "class": dotClassName,
      "style": {
        borderColor: /blue|red|green|gray/.test(color) ? undefined : color
      }
    }, [dot]), (0, _vue.createVNode)("div", {
      "class": "".concat(prefixCls, "-item-content")
    }, [this.$slots.default && this.$slots.default()])]);
  }
};
exports.default = _default;
