// ant-design-vue/lib/vc-menu/commonPropsType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  prefixCls: _vueTypes.default.string.def('rc-menu'),
  focusable: _vueTypes.default.bool.def(true),
  multiple: _vueTypes.default.bool,
  defaultActiveFirst: _vueTypes.default.bool,
  visible: _vueTypes.default.bool.def(true),
  activeKey: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]),
  selectedKeys: _vueTypes.default.arrayOf(_vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number])),
  defaultSelectedKeys: _vueTypes.default.arrayOf(_vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number])).def([]),
  defaultOpenKeys: _vueTypes.default.arrayOf(_vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number])).def([]),
  openKeys: _vueTypes.default.arrayOf(_vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number])),
  openAnimation: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.object]),
  mode: _vueTypes.default.oneOf(['horizontal', 'vertical', 'vertical-left', 'vertical-right', 'inline']).def('vertical'),
  triggerSubMenuAction: _vueTypes.default.string.def('hover'),
  subMenuOpenDelay: _vueTypes.default.number.def(0.1),
  subMenuCloseDelay: _vueTypes.default.number.def(0.1),
  level: _vueTypes.default.number.def(1),
  inlineIndent: _vueTypes.default.number.def(24),
  theme: _vueTypes.default.oneOf(['light', 'dark']).def('light'),
  getPopupContainer: _vueTypes.default.func,
  openTransitionName: _vueTypes.default.string,
  forceSubMenuRender: _vueTypes.default.bool,
  selectable: _vueTypes.default.bool,
  isRootMenu: _vueTypes.default.bool.def(true),
  builtinPlacements: _vueTypes.default.object.def(function () {
    return {};
  }),
  itemIcon: _vueTypes.default.any,
  expandIcon: _vueTypes.default.any,
  overflowedIndicator: _vueTypes.default.any
};
exports.default = _default;
