// ant-design-vue/lib/input/inputProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  prefixCls: _vueTypes.default.string,
  inputPrefixCls: _vueTypes.default.string,
  defaultValue: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]),
  value: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]),
  placeholder: [String, Number],
  type: {
    default: 'text',
    type: String
  },
  name: String,
  size: _vueTypes.default.oneOf(['small', 'large', 'default']),
  disabled: _vueTypes.default.bool,
  readonly: _vueTypes.default.bool,
  addonBefore: _vueTypes.default.any,
  addonAfter: _vueTypes.default.any,
  // onPressEnter?: React.FormEventHandler<any>;
  // onKeyDown?: React.FormEventHandler<any>;
  // onChange?: React.ChangeEventHandler<HTMLInputElement>;
  // onClick?: React.FormEventHandler<any>;
  // onFocus?: React.FormEventHandler<any>;
  // onBlur?: React.FormEventHandler<any>;
  prefix: _vueTypes.default.any,
  suffix: _vueTypes.default.any,
  // spellCheck: Boolean,
  autofocus: Boolean,
  allowClear: Boolean,
  lazy: {
    default: true,
    type: Boolean
  },
  maxlength: _vueTypes.default.number,
  loading: _vueTypes.default.bool,
  onPressEnter: _vueTypes.default.func,
  onKeydown: _vueTypes.default.func,
  onKeyup: _vueTypes.default.func,
  onFocus: _vueTypes.default.func,
  onBlur: _vueTypes.default.func,
  onChange: _vueTypes.default.func,
  onInput: _vueTypes.default.func,
  'onUpdate:value': _vueTypes.default.func
};
exports.default = _default;
