// ant-design-vue/lib/vc-collapse/src/commonProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.panelProps = exports.collapseProps = void 0;

var _vueTypes = _interopRequireDefault(require("../../_util/vue-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var collapseProps = function collapseProps() {
  return {
    prefixCls: _vueTypes.default.string,
    activeKey: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number, _vueTypes.default.arrayOf(_vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]))]),
    defaultActiveKey: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number, _vueTypes.default.arrayOf(_vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]))]),
    accordion: _vueTypes.default.bool,
    destroyInactivePanel: _vueTypes.default.bool,
    bordered: _vueTypes.default.bool,
    expandIcon: _vueTypes.default.func,
    openAnimation: _vueTypes.default.object,
    expandIconPosition: _vueTypes.default.oneOf(['left', 'right']),
    'onUpdate:change': _vueTypes.default.func,
    onChange: _vueTypes.default.func
  };
};

exports.collapseProps = collapseProps;

var panelProps = function panelProps() {
  return {
    openAnimation: _vueTypes.default.object,
    prefixCls: _vueTypes.default.string,
    header: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number, _vueTypes.default.node]),
    headerClass: _vueTypes.default.string,
    showArrow: _vueTypes.default.bool,
    isActive: _vueTypes.default.bool,
    destroyInactivePanel: _vueTypes.default.bool,
    disabled: _vueTypes.default.bool,
    accordion: _vueTypes.default.bool,
    forceRender: _vueTypes.default.bool,
    expandIcon: _vueTypes.default.func,
    extra: _vueTypes.default.any,
    panelKey: _vueTypes.default.any
  };
};

exports.panelProps = panelProps;
